server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceComputer.java

```diff
@@ -113,12 +118,21 @@ public DesiredBalance compute(
             logger.debug("Recomputing desired balance for [{}]", desiredBalanceInput.index());
         }
 
+        // A new ClusterInfo has arrived, clear the tracking for started shards
+        if (lastClusterInfo != desiredBalanceInput.routingAllocation().clusterInfo()) {
+            lastClusterInfo = desiredBalanceInput.routingAllocation().clusterInfo();
+            shardsStartedByAllocate = new HashMap<>();
+        }
+
+        final var alreadySimulatedStartedShards = new HashSet<ShardForSimulation>();
```

Comment:
Could we ask the simulator for the "alreadySimulatedShards" instead of tracking them alongside the simulateShardStarted calls? or am I missing something?

Reply:
Do you mean tracking it inside `ClusterInfoSimulator`? Yes that's possible. It does not do that currently but we can make it do so.

Reply:
Yeah that's what I was thinking, assume someone adds a new `simulateShardStarted` call, they'd have to remember to also update `alreadySimulatedShards`, but if we put it in `ClusterInfoSimulator` it would all be centralised.